sql: TestValidationWithProtectedTS failed

sql.TestValidationWithProtectedTS [failed](https://tanzanite.cluster.engflow.com/invocations/default/4c23641d-34a9-4829-8f4d-d95162b656df?testReportRun=7&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [80c77c6651754d31dfb2c852de419d032e0d412b](https://github.com/cockroachdb/cockroach/commits/80c77c6651754d31dfb2c852de419d032e0d412b):

Fatal error:

```
panic: test timed out after 14m57s
running tests:
	TestValidationWithProtectedTS (14m56s)

```
Stack: 

```
goroutine 807742 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestValidationWithProtectedTS
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestValidationWithProtectedTS3301513052
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:144: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=7</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestValidationWithProtectedTS.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-38354

--- a/pkg/sql/backfill_protected_timestamp_test.go
+++ b/pkg/sql/backfill_protected_timestamp_test.go
@@ -136,6 +136,11 @@ func TestValidationWithProtectedTS(t *testing.T) {
 	grp := ctxgroup.WithContext(ctx)
 	grp.Go(func() error {
 		<-indexValidationQueryWait
+		defer func() {
+			// Always unblock the validation query, even if something
+			// fails here.
+			indexValidationQueryResume <- struct{}{}
+		}()
 		getTableRangeIDs := func(t *testing.T) ([]int64, error) {
 			t.Helper()
 			rows, err := db.QueryContext(ctx, "WITH r AS (SHOW RANGES FROM TABLE t) SELECT range_id FROM r ORDER BY start_key")
@@ -160,6 +165,7 @@ func TestValidationWithProtectedTS(t *testing.T) {
 			return err
 		}
 		const retryTxnErrorSubstring = "restart transaction"
+		const replicaGCError = "must be after replica GC threshold"
 		for {
 			if _, err := db.ExecContext(ctx, "BEGIN"); err != nil {
 				return err
@@ -175,7 +181,8 @@ func TestValidationWithProtectedTS(t *testing.T) {
 			}
 			_, err = db.ExecContext(ctx, "COMMIT")
 			if err != nil {
-				if strings.Contains(err.Error(), retryTxnErrorSubstring) {
+				if strings.Contains(err.Error(), retryTxnErrorSubstring) ||
+					strings.Contains(err.Error(), replicaGCError) {
 					err = nil
 					continue
 				}
@@ -190,7 +197,6 @@ func TestValidationWithProtectedTS(t *testing.T) {
 				return err
 			}
 		}
-		indexValidationQueryResume <- struct{}{}
 		return nil
 	})
 	grp.Go(func() error {
@@ -199,12 +205,21 @@ func TestValidationWithProtectedTS(t *testing.T) {
 	})
 
 	require.NoError(t, grp.Wait())
-	// Validate the rows were removed due to the drop
-	res := r.QueryStr(t, `SELECT n FROM t@foo`)
+	// Validate the rows were removed due to the delete.
+	// Note: Because of the low GC timestamp we can hit "must be after replica GC
+	// threshold" errors.
+	var res [][]string
+	testutils.SucceedsSoon(t, func() error {
+		rows, err := r.DB.QueryContext(ctx, `SELECT n FROM t@foo`)
+		if err != nil {
+			return err
+		}
+		res, err = sqlutils.RowsToStrMatrix(rows)
+		return err
+	})
 	if len(res) != 1 {
 		t.Errorf("expected %d entries, got %d", 1, len(res))
 	}
-	require.NoError(t, db.Close())
 }
 
 // TestBackfillQueryWithProtectedTS backfills a query into a table and confirms

